tasmota/xdrv_38_somfy.ino

```diff
@@ -0,0 +1,198 @@
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef USE_SOMFY
+
+#define XDRV_38 38
+
+/*********************************************************************************************\
+ * Send Somfy RTS commands via a connected 433MHz transmitter module
+\*********************************************************************************************/
+const char kSomfyCommands[] PROGMEM = "Somfy|" // prefix
+  "Up|Down|Stop|Prog|Set";
+
+void (* const somfyCommand[])(void) PROGMEM = {
+  &CmdUp, &CmdDown, &CmdStop, &CmdProg, &CmdSomfySet};
+
+#define SYMBOL 640
+#define SOMFY_UP 0x2
+#define SOMFY_STOP 0x1
+#define SOMFY_DOWN 0x4
+#define SOMFY_PROG 0x8
+
+byte frame[7];
+byte checksum;
+
+void BuildFrame(byte *frame, byte button);
+void SendCommand(byte *frame, byte sync);
+
+void somfysetup() {
+  pinMode(pin[GPIO_SOMFY_RTS_TX], OUTPUT);
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], LOW);
+}
+
+void BuildFrame(byte *frame, byte button) {
+  unsigned int code, remote;
+  code = Settings.rolling_code_count;
+  remote = Settings.rolling_code_serial;
+  frame[0] = 0xA7; // Encryption key. Doesn't matter much
+  frame[1] = button << 4;  // Which button did you press? The 4 LSB will be the checksum
+  frame[2] = code >> 8;    // Rolling code (big endian)
+  frame[3] = code;         // Rolling code
+  frame[4] = remote >> 16; // Remote address
+  frame[5] = remote >>  8; // Remote address
+  frame[6] = remote;       // Remote address
+// Checksum calculation: a XOR of all the nibbles
+  checksum = 0;
+  for(byte i = 0; i < 7; i++) {
+    checksum = checksum ^ frame[i] ^ (frame[i] >> 4);
+  }
+  checksum &= 0b1111; // We keep the last 4 bits only
+//Checksum integration
+  frame[1] |= checksum; //  If a XOR of all the nibbles is equal to 0, the blinds will
+                        // consider the checksum ok.
+// Obfuscation: a XOR of all the bytes
+  for(byte i = 1; i < 7; i++) {
+    frame[i] ^= frame[i-1];
+  }
+
+  code++;
+  Settings.rolling_code_count = code;
+}
+
+void SendCommand(byte *frame, byte sync) {
+  if(sync == 2) { // Only with the first frame.
+  //Wake-up pulse & Silence
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], HIGH);
+    delayMicroseconds(9415);
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], LOW);
+    delayMicroseconds(89565);
+  }
+// Hardware sync: two sync for the first frame, seven for the following ones.
+  for (int i = 0; i < sync; i++) {
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], HIGH);
+    delayMicroseconds(4*SYMBOL);
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], LOW);
+    delayMicroseconds(4*SYMBOL);
+  }
+// Software sync
+  digitalWrite(pin[GPIO_SOMFY_RTS_TX], HIGH);
+  delayMicroseconds(4550);
```

Comment:
is there any reason why you do not use here a different value instead of using ``SYMBOL`` as well?

Reply:
Hi, I was limiting refactoring to make the v 4.4 driver work with the current version, but I am not the original author. See corresponding issue. There are still   lots of improvements possible 😉